design/ceph-csi-operator.md

```diff
@@ -0,0 +1,84 @@
+---
+title: Integrate cephcsi operator with Rook
+target-version: release-1.15
+---
+
+# Integrate cephcsi operator with Rook
+
+## Overview
+
+To integrate the new ceph-csi operator with Rook and remove the existing csi deployment related code, RBAC's and dead code in later releases. Also before moving forward please read the ceph-csi-operator design doc [here](https://github.com/ceph/ceph-csi-operator/blob/main/docs/design/operator.md).
+
+## Rook Deploying Ceph-CSI-Operator
+
+There are two scenarios rook needs to manage while deploying of the new csi-operator.
+
+1. Deployment in new cluster i.e [Fresh deployment](#fresh-deployment)
+    1. Manifest Deployment
+    2. Helm Deployment
+2. Deployment in upgrade cluster i.e [Upgrade clusters](#upgraded-clusters) cluster.
+    1. Manifest Upgrade
+    2. Helm Upgrade
+
+We need to have some bash script which will be called by some make command to automatically update the RBAC, CRDs files in rook repo from the ceph-csi-operator repo.
+
+## Fresh deployment
+
+Rook will define the new csi-operator defaults settings in [CRDs](#new-ceph-csi-operator-crds) in operator.yaml or operator chart `rook-ceph` `value.yaml`, and can be customizable by users. Rook will define the CSI operator and related CRs such as it is deployed by default and automatically with Rook.
+
+- **Manifests**:
+    - operator.yaml: Include the CSI operator deployment, and default CephCSI CRs
+    - crds.yaml: Include the CephCSI CRDs
+    - common.yaml: Include the RBAC needed for the CSI operator
+    - Consider whether to have separate manifests, particularly in the first release if the CSI operator is experimental
+- **Helm chart**: Add new settings in the operator and cluster chart and all others CRDs, RBAC will be added similarly to the manifest
+
+In fresh clusters, we will introduce a new setting in `operator.yaml` configmap `rook-ceph-operator-config` name `ROOK_USE_CSI_OPERATOR: true` and if `USE_CSI_OPERATOR:` is set to `true` it will also disable the csi `ROOK_CSI_DISABLE_DRIVER`. Setting `ROOK_CSI_DISABLE_DRIVER: true` will skip the current csi driver reconciliation and the settings configured in `operator.yaml` for the new CRs will be taken into consideration.
+
+From here we can proceed two ways to configure other settings and deploy the csi-operator.
+
+1. Rook create the default manifest present in rook example directory.
+2. User/Admin creates/modify the default manifest itself present in rook repository
+
+## Upgraded clusters
+
+For upgrade clusters, we'll fetch the csi settings and cluster connections details from configmaps like `rook-ceph-operator-config` and `rook-ceph-csi-config`. After fetching the csi settings and cluster connections
```

Comment:
What is the impact to users? We expect that this will be automatic and no impact to users with their existing storage classes, right?

Reply:
And the user could switch back and forth between the old and new CSI settings, right? For example, if they enable the experimental CSI operator and there is some issue, they could revert back to the previous CSI driver?

Reply:
But after we decide it is no longer experimental, we could just enforce conversion to the CSI operator and completely remove support for the old one.

Reply:
> What is the impact of upgrades to users? We expect that this will be automatic and no impact to users with their existing storage classes, right?

Yes, this is our goal to have smooth upgrade

Reply:
> And the user could switch back and forth between the old and new CSI settings, right? For example, if they enable the experimental CSI operator and there is some issue, they could revert back to the previous CSI driver?

I'm not sure, if users will able to do back and forth. Switching from the current to the new way of CSI deployment is the goal. cc @Madhu-1 comments

Reply:
To be clear, smooth/automated upgrades can come in stable mode, not needed for experimental.

Reply:
Yes this will be only for stable mode to fallback if something is not working as expected